raiden/raiden_service.py

```diff
@@ -210,108 +283,266 @@ def register_secret(self, secret):
         and ignoring the tokens. Useful for refund transfer, split transfer,
         and exchanges.
         """
-        for token_manager in self.managers_by_token_address.values():
-            try:
-                token_manager.register_secret(secret)
-            except:  # pylint: disable=bare-except
-                # Only channels that care about the given secret can be
-                # registered and channels that have claimed the lock must
-                # be removed, so an exception should not happen at this
-                # point, nevertheless handle it because we dont want a
-                # error in a channel to mess the state from others.
-                log.error('programming error')
+        hashlock = sha3(secret)
+        revealsecret_message = RevealSecret(secret)
+        self.sign(revealsecret_message)
 
-    def message_for_task(self, message, hashlock):
-        """ Sends the message to the corresponding task.
+        for hash_channel in self.hashlock_channel.itervalues():
+            for channel in hash_channel[hashlock]:
+                try:
+                    channel.register_secret(secret)
+
+                    # This will potentially be executed multiple times and could suffer
+                    # from amplification, the protocol will ignore messages that were
```

Comment:
I might have been over pessimist with this, but here is the issue: https://github.com/raiden-network/raiden/issues/493